src/main/java/htsjdk/samtools/util/IntervalList.java

```diff
@@ -869,4 +844,68 @@ public int hashCode() {
         result = 31 * result + intervals.hashCode();
         return result;
     }
+
+    public static class IntervalMerger implements Iterable<Interval> {
+
+        Iterator<Interval> inputIntervals;
+        Interval current = null;
+        final boolean combineAbuttingIntervals;
+        final boolean enforceSameStrands;
+        final boolean concatenateNames;
+        final List<Interval> toBeMerged = new ArrayList<>();
+
+        public IntervalMerger(Iterator<Interval> intervals, final boolean combineAbuttingIntervals, final boolean enforceSameStrand, final boolean concatenateNames) {
+            this.inputIntervals = intervals;
+
+            this.combineAbuttingIntervals = combineAbuttingIntervals;
+            this.enforceSameStrands = enforceSameStrand;
+            this.concatenateNames = concatenateNames;
+        }
+
+        @Override
+        public Iterator<Interval> iterator() {
+
+            return new Iterator<Interval>() {
+                @Override
+                public boolean hasNext() {
+                    return current != null || inputIntervals.hasNext();
+                }
+
+                @Override
+                public Interval next() {
+                    return getNext();
+                }
+
+                private Interval getNext() {
+                    Interval next = null;
+                    while (inputIntervals.hasNext()) {
+                        next = inputIntervals.next();
+                        if (current == null) {
+                            toBeMerged.add(next);
+                            current = next;
+                        } else if (current.intersects(next) || (combineAbuttingIntervals && current.abuts(next))) {
+                            if (enforceSameStrands && current.isNegativeStrand() != next.isNegativeStrand()) {
+                                throw new SAMException("Strands were not equal for: " + current.toString() + " and " + next.toString());
+                            }
+                            toBeMerged.add(next);
+                            current = new Interval(current.getContig(), current.getStart(), Math.max(current.getEnd(), next.getEnd()), current.isNegativeStrand(), null);
```

Comment:
It would be more efficient to have a mutable object that we just update with the new bounds instead of creating a new object each round, but that's probably a micro-optimization that isn't worth bothering with...

Reply:
added a MutableFeature for this purpose....